Implement WebProcess side of InstallEvent addRoutes

youennf opened the pull request:
#### b949f0f6d399dd158150f29b298f930a02d43d08
<pre>
Implement WebProcess side of InstallEvent addRoutes
<a href="https://rdar.apple.com/144078980">rdar://144078980</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=286914">https://bugs.webkit.org/show_bug.cgi?id=286914</a>

Reviewed by Ben Nham.

Add parsing of service worker routes and handling in service worker context.
The data is sent up to WebSWClientConnection which rejects the promise (next patch will implement this method).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.css: Added.
(body):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/imported-sw.js: Added.
(key.async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct1.text: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct1.text.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct2.text: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct2.text.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/router-rules.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-helpers.sub.js: Added.
(const.get_info_from_worker.async worker.):
(const.get_info_from_worker.async worker):
(const.reset_info_in_worker.async worker.):
(const.reset_info_in_worker.async worker):
(const.scriptDir.document.currentScript.src.match.const.registerAndActivate.async test):
(const.createIframe.async t):
(name.return.promise_test.async t):
(iframeTest):
(randomString):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-no-fetch-handler-sw.js: Added.
(async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-race-network-and-fetch-handler-sw.js: Added.
(async e):
(url.searchParams.has):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-sw.js: Added.
(async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-sw.sub.js: Added.
(const.recordError):
(const.getRecords):
(const.resetRecords):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-fetch-event.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-fetch-event.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-invalid-rules.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-invalid-rules.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-main-resource.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-main-resource.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-multiple-router-registrations.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-multiple-router-registrations.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-mutiple-conditions.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-mutiple-conditions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-no-fetch-handler.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-no-fetch-handler.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-race-network-and-fetch-handler.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-race-network-and-fetch-handler.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-destination.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-destination.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-method.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-method.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-subresource.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-subresource.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/static-router-resource-timing.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/static-router-resource-timing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/w3c-import.log: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::addExtendLifetimePromise):
(WebCore::ExtendableEvent::whenAllExtendLifetimePromisesAreSettled):
* Source/WebCore/workers/service/ExtendableEvent.h:
(WebCore::ExtendableEvent::isWaiting const):
* Source/WebCore/workers/service/InstallEvent.cpp:
(WebCore::toServiceWorkerRoutePattern):
(WebCore::toServiceWorkerRouteCondition):
(WebCore::toServiceWorkerRoute):
(WebCore::verifyRouterCondition):
(WebCore::addServiceWorkerRoute):
(WebCore::InstallEvent::addRoutes):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContextData.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::storeEventTypesToHandle):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRoute.cpp: Added.
(WebCore::validateURLPatternComponent):
(WebCore::validateServiceWorkerRouteCondition):
(WebCore::validateServiceWorkerRoute):
(WebCore::matchURLPatternComponent):
(WebCore::matchURLPattern):
(WebCore::matchRouterCondition):
(WebCore::ServiceWorkerRouteCondition::isolatedCopy):
(WebCore::ServiceWorkerRouteCondition::copy const):
(WebCore::ServiceWorkerRoutePattern::isolatedCopy):
(WebCore::crossThreadCopyRouterSource):
(WebCore::ServiceWorkerRoute::isolatedCopy):
* Source/WebCore/workers/service/ServiceWorkerRoute.h: Added.
(WebCore::ServiceWorkerRoute::copy const):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::addRoutes):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireInstallEvent):
(WebCore::ServiceWorkerThread::finishedEvaluatingScript):
(WebCore::createInstallEvent): Deleted.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::addRoutes):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:

Canonical link: <a href="https://commits.webkit.org/291492@main">https://commits.webkit.org/291492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87e275aecfa399574a2cfb5b3ecf645df7479f8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93059 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98058 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43585 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95109 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21065 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71154 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28575 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96061 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9719 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51482 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9413 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1843 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42898 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79704 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100084 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20111 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80178 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80107 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79480 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24025 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1339 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13191 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20095 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25271 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19782 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21523 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebCore/workers/service/ServiceWorkerRoute.cpp:
```diff
@@ -0,0 +1,261 @@
+{
+    std::unique_ptr<ServiceWorkerRouteCondition> notConditionCopy;
+    if (notCondition)
+        notConditionCopy = makeUnique<ServiceWorkerRouteCondition>(WTFMove(*notCondition));
```
How does moving `*notCondition` even work in this const member function? Maybe I am misreading it. Is it just because an rvalue reference is allowed to bind to a const lvalue reference?

youennf replied:
WTFMove is indeed wrong, I'll remove it.

youennf committed:
Implement WebProcess side of InstallEvent addRoutes
rdar://144078980
https://bugs.webkit.org/show_bug.cgi?id=286914

Reviewed by Ben Nham.

Add parsing of service worker routes and handling in service worker context.
The data is sent up to WebSWClientConnection which rejects the promise (next patch will implement this method).

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.css: Added.
(body):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/direct.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/imported-sw.js: Added.
(key.async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct1.text: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct1.text.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct2.text: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/direct2.text.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/or-test/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/router-rules.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-helpers.sub.js: Added.
(const.get_info_from_worker.async worker.):
(const.get_info_from_worker.async worker):
(const.reset_info_in_worker.async worker.):
(const.reset_info_in_worker.async worker):
(const.scriptDir.document.currentScript.src.match.const.registerAndActivate.async test):
(const.createIframe.async t):
(name.return.promise_test.async t):
(iframeTest):
(randomString):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-no-fetch-handler-sw.js: Added.
(async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-race-network-and-fetch-handler-sw.js: Added.
(async e):
(url.searchParams.has):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-sw.js: Added.
(async e):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/static-router-sw.sub.js: Added.
(const.recordError):
(const.getRecords):
(const.resetRecords):
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-fetch-event.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-fetch-event.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-invalid-rules.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-invalid-rules.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-main-resource.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-main-resource.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-multiple-router-registrations.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-multiple-router-registrations.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-mutiple-conditions.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-mutiple-conditions.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-no-fetch-handler.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-no-fetch-handler.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-race-network-and-fetch-handler.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-race-network-and-fetch-handler.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-destination.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-destination.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-method.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-request-method.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-subresource.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/static-router-subresource.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/static-router-resource-timing.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/static-router-resource-timing.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/tentative/static-router/w3c-import.log: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::addExtendLifetimePromise):
(WebCore::ExtendableEvent::whenAllExtendLifetimePromisesAreSettled):
* Source/WebCore/workers/service/ExtendableEvent.h:
(WebCore::ExtendableEvent::isWaiting const):
* Source/WebCore/workers/service/InstallEvent.cpp:
(WebCore::toServiceWorkerRoutePattern):
(WebCore::toServiceWorkerRouteCondition):
(WebCore::toServiceWorkerRoute):
(WebCore::verifyRouterCondition):
(WebCore::addServiceWorkerRoute):
(WebCore::InstallEvent::addRoutes):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContextData.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::storeEventTypesToHandle):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRoute.cpp: Added.
(WebCore::validateURLPatternComponent):
(WebCore::validateServiceWorkerRouteCondition):
(WebCore::validateServiceWorkerRoute):
(WebCore::matchURLPatternComponent):
(WebCore::matchURLPattern):
(WebCore::matchRouterCondition):
(WebCore::ServiceWorkerRouteCondition::isolatedCopy):
(WebCore::ServiceWorkerRouteCondition::copy const):
(WebCore::ServiceWorkerRoutePattern::isolatedCopy):
(WebCore::crossThreadCopyRouterSource):
(WebCore::ServiceWorkerRoute::isolatedCopy):
* Source/WebCore/workers/service/ServiceWorkerRoute.h: Added.
(WebCore::ServiceWorkerRoute::copy const):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::addRoutes):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireInstallEvent):
(WebCore::ServiceWorkerThread::finishedEvaluatingScript):
(WebCore::createInstallEvent): Deleted.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::addRoutes):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:

Canonical link: https://commits.webkit.org/291492@main
